File: prototype/node/src/app/core/orders/order-placement.ts
import type { ListingStatus } from '../listings/listing-status.ts'

export const UNAVAILABLE_REASONS = ['removed', 'off_sale', 'sold_out', 'short_stock'] as const
export type UnavailableReason = (typeof UNAVAILABLE_REASONS)[number]

/** A cart line as placement judges it: what the cart holds, against what the
 * listing is now. */
export type PlaceableLine = {
  listingId: number
  title: string
  status: ListingStatus
  availableQuantity: number
  quantity: number
  hasActiveRemoval: boolean
}

export type UnavailableLine = {
  listingId: number
  title: string
  reason: UnavailableReason
}

/**
 * Whether a cart may become an order, and what stands in the way when it may
 * not. Generic in the line so the caller keeps whatever else it read alongside
 * the fields placement looks at.
 */
export type OrderPlacement<Line extends PlaceableLine> =
  | { ok: true; lines: readonly Line[] }
  | { ok: false; unavailable: readonly UnavailableLine[] }

function unavailableReason(line: PlaceableLine): UnavailableReason | null {
  if (line.hasActiveRemoval) return 'removed'
  if (line.status === 'sold') return 'sold_out'
  if (line.status !== 'for_sale') return 'off_sale'
  if (line.availableQuantity < 1) return 'sold_out'
  if (line.quantity > line.availableQuantity) return 'short_stock'

  return null
}

/** Every line that stands in the way is named, so one answer covers the cart. */
export function planOrderPlacement<Line extends PlaceableLine>(
  lines: readonly Line[],
): OrderPlacement<Line> {
  const unavailable = lines.flatMap((line) => {
    const reason = unavailableReason(line)

    return reason === null ? [] : [{ listingId: line.listingId, title: line.title, reason }]
  })

  return unavailable.length === 0 ? { ok: true, lines } : { ok: false, unavailable }
}

export type UnavailableNotice = {
  title: string
  notice: string
}

const UNAVAILABLE_NOTICES: Readonly<Record<UnavailableReason, string>> = {
  removed: 'no longer available',
  off_sale: 'no longer for sale',
  sold_out: 'sold out',
  short_stock: 'no longer in stock in that quantity',
}

/** What the checkout page says about each line it cannot buy. */
export function unavailableNotices(
  lines: readonly UnavailableLine[],
): readonly UnavailableNotice[] {
  return lines.map((line) => ({ title: line.title, notice: UNAVAILABLE_NOTICES[line.reason] }))
}

File: prototype/rails/src/app/models/order_placement.rb
# Whether a cart or order may become (or stay) a placed sale, and which lines
# stand in the way when it may not. Built from anything that responds to
# `listing` and `quantity` — a cart item at checkout, an order item at a
# retried charge — so one classifier serves both callers.
class OrderPlacement
  Line = Data.define(:listing_id, :title, :status, :available_quantity, :quantity, :removed)
  BlockedLine = Data.define(:listing_id, :title, :reason)

  NOTICES = {
    removed: "no longer available",
    off_sale: "no longer for sale",
    sold_out: "sold out",
    short_stock: "no longer in stock in that quantity"
  }.freeze

  class << self
    def plan(items)
      new(items)
    end

    # Which reason blocks a line, or nil when nothing does. A removal
    # outranks whatever the status says; sold outranks a bare quantity of
    # zero reading as merely out of stock.
    def reason_for(line)
      return :removed if line.removed
      return :sold_out if line.status == "sold"
      return :off_sale if line.status != "for_sale"
      return :sold_out if line.available_quantity < 1
      :short_stock if line.quantity > line.available_quantity
    end

    def notice_for(reason)
      NOTICES.fetch(reason)
    end

    # The blocked lines' listing id, title, and reason, shaped for a log line.
    def log_payload(blocked_lines)
      blocked_lines.map { |line| { listing_id: line.listing_id, title: line.title, reason: line.reason } }
    end

    def line_for(item)
      listing = item.listing

      Line.new(
        listing_id: listing.id, title: listing.title, status: listing.status,
        available_quantity: listing.quantity, quantity: item.quantity,
        removed: listing.actively_removed?
      )
    end
  end

  attr_reader :items

  def initialize(items)
    @items = items
  end

  def ok?
    blocked_lines.empty?
  end

  def blocked_lines
    @blocked_lines ||= items.filter_map { |item| blocked_line(item) }
  end

  private

  def blocked_line(item)
    line = self.class.line_for(item)
    reason = self.class.reason_for(line)
    return nil if reason.nil?

    BlockedLine.new(listing_id: line.listing_id, title: line.title, reason: reason)
  end
end
